### PR TITLE
fix: panic of start leading. sync ingress failed when apisix not start.

### PR DESCRIPTION
**Pull request by slene**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues

#448 
#449

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: pkg/apisix/apisix.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ import (
 	"sync"
 
 	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
```
Comment (tokers) [inline pkg/apisix/apisix.go:22]: By convention, we put the 3-party package at the middle of `import`.

```go
import (
	std
	......

	"k8s.io/apimachinery/pkg/util/wait"

	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
)
```

File: pkg/apisix/cluster.go [hunk 2]
```diff
@@ -50,6 +50,19 @@ var (
 	ErrDuplicatedCluster = errors.New("duplicated cluster")
 
 	_errReadOnClosedResBody = errors.New("http: read on closed response body")
+
+	// Default shared transport if apisix client
+	defaultTransport = &http.Transport{
```
Comment (tokers) [inline pkg/apisix/cluster.go:55]:
```suggestion
	_defaultTransport = &http.Transport{
```

File: pkg/apisix/cluster.go [hunk 3]
```diff
@@ -290,6 +303,39 @@ func (c *cluster) GlobalRule() GlobalRule {
...
+	})
+	if err != nil {
+		// if ErrWaitTimeout then set lastSyncErr
+		c.cacheSyncErr = lastCheckErr
+	}
+	return err
+}
+
+func (c *cluster) healthCheck(ctx context.Context) (err error) {
+	// TODO
```
Comment (tokers) [inline pkg/apisix/cluster.go:332]: Just use TCP socket probe is OK.

File: pkg/apisix/nonexistentclient.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import (
 
 	"github.com/apache/apisix-ingress-controller/pkg/apisix/cache"
 	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
```
Comment (tokers) [inline pkg/apisix/nonexistentclient.go:23]: Ditto with the `import` convention.

File: pkg/config/config.go [hunk 5]
```diff
@@ -85,6 +85,8 @@ type APISIXConfig struct {
 	// DefaultClusterAdminKey is the admin key for the default cluster.
 	// TODO: Obsolete the plain way to specify admin_key, which is insecure.
 	DefaultClusterAdminKey string `json:"default_cluster_admin_key" yaml:"default_cluster_admin_key"`
+	// DefaultClusterClientTimeout is the request timeout for default cluster client.
+	DefaultClusterClientTimeout types.TimeDuration `json:"default_cluster_client_timeout" yaml:"default_cluster_client_timeout"`
```
Comment (tokers) [inline pkg/config/config.go:89]: We will introduce `ApisixClusterConfig` to set these options, so don't add it here.

File: pkg/ingress/apisix_cluster_config.go [hunk 6]
```diff
@@ -62,7 +64,6 @@ func (c *apisixClusterConfigController) run(ctx context.Context) {
 		go c.runWorker(ctx)
 	}
 	<-ctx.Done()
-	c.workqueue.ShutDown()
```
Comment (tokers) [inline pkg/ingress/apisix_cluster_config.go:65]: It's not related to PR change. If you wanna change them to use `defer`, just in another PR.

File: pkg/ingress/controller.go [hunk 7]
```diff
@@ -317,30 +326,42 @@ election:
 }
 
 func (c *Controller) run(ctx context.Context) {
-	log.Infow("controller now is running as leader",
+	log.Infow("controller is start leading ...",
```
Comment (tokers) [inline pkg/ingress/controller.go:329]:
```suggestion
	log.Infow("controller tries to leading ...",
```

File: pkg/ingress/controller.go [hunk 8]
```diff
@@ -317,30 +326,42 @@ election:
...
+	clusterOpts := &apisix.ClusterOptions{
 		Name:     c.cfg.APISIX.DefaultClusterName,
 		AdminKey: c.cfg.APISIX.DefaultClusterAdminKey,
 		BaseURL:  c.cfg.APISIX.DefaultClusterBaseURL,
-	})
+		Timeout:  c.cfg.APISIX.DefaultClusterClientTimeout.Duration,
+	}
+	err := c.apisix.AddCluster(clusterOpts)
 	if err != nil && err != apisix.ErrDuplicatedCluster {
-		// TODO give up the leader role.
```
Comment (tokers) [inline pkg/ingress/controller.go:333]: Don't remove the TODO, unless you really achieve this.

File: pkg/ingress/controller.go [hunk 9]
```diff
@@ -317,30 +326,42 @@ election:
...
 		log.Errorf("failed to add default cluster: %s", err)
 		return
 	}
 
 	if err := c.apisix.Cluster(c.cfg.APISIX.DefaultClusterName).HasSynced(ctx); err != nil {
-		// TODO give up the leader role.
```
Comment (tokers) [inline pkg/ingress/controller.go:339]: Ditto.

File: pkg/apisix/apisix.go [hunk 10]
```diff
@@ -50,6 +51,8 @@ type Cluster interface {
 	String() string
 	// HasSynced checks whether all resources in APISIX cluster is synced to cache.
 	HasSynced(context.Context) error
+	// HealthCheck checks apisix cluster health in realtime.
+	HealthCheck(context.Context, wait.Backoff) error
```
Comment (tokers) [inline pkg/apisix/apisix.go:55]: IMHO We shall not accept a `wait.Backoff` object from the caller. What we try to do is hide some details and just tell the caller whether the cluster is healthy or not.

File: pkg/ingress/controller.go [hunk 11]
```diff
@@ -180,40 +158,71 @@ func NewController(cfg *config.Config) (*Controller, error) {
...
-		apisixUpstreamLister:        kubeClient.APISIXSharedIndexInformerFactory.Apisix().V1().ApisixUpstreams().Lister(),
-		apisixTlsInformer:           kubeClient.APISIXSharedIndexInformerFactory.Apisix().V1().ApisixTlses().Informer(),
-		apisixTlsLister:             kubeClient.APISIXSharedIndexInformerFactory.Apisix().V1().ApisixTlses().Lister(),
-		apisixClusterConfigInformer: kubeClient.APISIXSharedIndexInformerFactory.Apisix().V2alpha1().ApisixClusterConfigs().Informer(),
-		apisixClusterConfigLister:   kubeClient.APISIXSharedIndexInformerFactory.Apisix().V2alpha1().ApisixClusterConfigs().Lister(),
 	}
+	return c, nil
+}
+
+func (c *Controller) initWhenStartLeader() {
```
Comment (tokers) [inline pkg/ingress/controller.go:165]:
```suggestion
func (c *Controller) initWhenStartLeading() {
```

File: pkg/kube/apisix/apis/config/v2alpha1/types.go [hunk 12]
```diff
@@ -17,6 +17,7 @@ package v2alpha1
 import (
 	"encoding/json"
 
+	"github.com/apache/apisix-ingress-controller/pkg/types"
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2alpha1/types.go:20]: Put our own packages at the bottom of `import`.

File: pkg/kube/init.go [hunk 13]
```diff
@@ -25,16 +25,12 @@ import (
 
 // KubeClient contains some objects used to communicate with Kubernetes API Server.
 type KubeClient struct {
+	cfg *config.Config
+
 	// Client is the object used to operate Kubernetes builtin resources.
 	Client kubernetes.Interface
 	// APISIXClient is the object used to operate resources under apisix.apache.org group.
 	APISIXClient clientset.Interface
-	// SharedIndexInformerFactory is the index informer factory object used to watch and
```
Comment (tokers) [inline pkg/kube/init.go:32]: Why change this?

File: pkg/ingress/controller.go [hunk 14]
```diff
@@ -423,3 +449,26 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
...
+		backoff := wait.Backoff{
+			Duration: 5 * time.Second,
+			Factor:   1,
+			Steps:    3,
+		}
+		err := c.apisix.Cluster(c.cfg.APISIX.DefaultClusterName).HealthCheck(ctx, backoff)
+		if err != nil {
+			// Finally failed health check, then give up leader.
+			c.leaderContextCancelFunc()
+			log.Warnf("failed to HealthCheck for default cluster: %s, give up leader", err)
```
Comment (tokers) [inline pkg/ingress/controller.go:470]:
```suggestion
			log.Warnf("failed to check health for default cluster: %s, give up leader", err)
```

File: pkg/ingress/controller.go [hunk 15]
```diff
@@ -423,3 +449,26 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
 	}
 	return err
 }
+
+func (c *Controller) checkClusterHealthy(ctx context.Context) {
```
Comment (tokers) [inline pkg/ingress/controller.go:453]:
```suggestion
func (c *Controller) checkClusterHealth(ctx context.Context) {
```

File: pkg/apisix/cluster.go (hunk 3 above)
Comment (slene) [reply pkg/apisix/cluster.go:332]: ok, later.

File: pkg/ingress/apisix_cluster_config.go (hunk 6 above)
Comment (slene) [reply pkg/ingress/apisix_cluster_config.go:65]: If not change them. Wiil cause goroutine leak in this pr. When re-create *controller.

File: pkg/ingress/controller.go (hunk 8 above)
Comment (slene) [reply pkg/ingress/controller.go:333]: Do heath check, give up leader if failed. Not enough ?

File: pkg/kube/init.go (hunk 13 above)
Comment (slene) [reply pkg/kube/init.go:32]:
```
// in controller
// re-execute this code will make panic `close of closed channel` when context cancelled
// unless you re-create Informer
**Informer.Run(ctx.Done())
```

So should re-create **Informer in every controller.run. It need re-create InformerFactory.

File: pkg/ingress/controller.go (hunk 8 above)
Comment (slene) [reply pkg/ingress/controller.go:333]: ok

File: pkg/ingress/controller.go [hunk 16]
```diff
@@ -423,3 +449,26 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
...
+	for {
+		select {
+		case <-ctx.Done():
+		case <-time.After(5 * time.Second):
+		}
+
+		// Retry three times in a row, and exit if all of them fail.
+		backoff := wait.Backoff{
```
Comment (gxthrj) [inline pkg/ingress/controller.go]: Put logic into the `case <-time.After(5 * time.Second)`

File: pkg/apisix/cluster.go [hunk 17]
```diff
@@ -132,13 +144,20 @@ func (c *cluster) syncCache() {
...
-		Steps:    6,
-	}
-	err := wait.ExponentialBackoff(backoff, c.syncCacheOnce)
+		Duration: 2 * time.Second,
+		Factor:   1,
+		Steps:    5,
+	}
+	var lastSyncErr error
+	err := wait.ExponentialBackoff(backoff, func() (done bool, _ error) {
+		// not possible return: false, nil
```
Comment (gxthrj) [inline pkg/apisix/cluster.go:153]:
```suggestion
		// impossibly return: false, nil
```

File: pkg/ingress/controller.go [hunk 18]
```diff
@@ -423,3 +455,21 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
...
+		select {
+		case <-ctx.Done():
+		case <-time.After(5 * time.Second):
+		}
+
+		err := c.apisix.Cluster(c.cfg.APISIX.DefaultClusterName).HealthCheck(ctx)
+		if err != nil {
+			// Finally failed health check, then give up leader.
+			log.Warnf("failed to check health for default cluster: %s, give up leader", err)
+			cancelFunc()
```
Comment (gxthrj) [inline pkg/ingress/controller.go:470]:
```suggestion
defer cancelFunc()
```

File: pkg/ingress/controller.go [hunk 19]
```diff
@@ -423,3 +455,21 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
...
+			return
```
Comment (gxthrj) [inline pkg/ingress/controller.go:471]: Should we set the state to `_cacheSyncing` before return?

File: pkg/ingress/apisix_cluster_config.go (hunk 6 above)
Comment (tokers) [reply pkg/ingress/apisix_cluster_config.go:65]: Did you mean the case that cache sync failure?

File: pkg/ingress/controller.go (hunk 19 above)
Comment (slene) [reply pkg/ingress/controller.go:471]: No, the current logic is to just discard it if it is unhealthy.

File: pkg/kube/init.go (hunk 13 above)
Comment (tokers) [reply pkg/kube/init.go:32]: OK, got it.

File: pkg/apisix/cluster.go [hunk 20]
```diff
@@ -50,6 +51,19 @@ var (
 	ErrDuplicatedCluster = errors.New("duplicated cluster")
 
 	_errReadOnClosedResBody = errors.New("http: read on closed response body")
+
+	// Default shared transport if apisix client
```
Comment (tokers) [inline pkg/apisix/cluster.go:55]:
```suggestion
	// Default shared transport for apisix client
```

We may export options to customize this transport in the future (in command line options).

File: pkg/apisix/cluster.go [hunk 21]
```diff
@@ -63,6 +77,7 @@ type ClusterOptions struct {
 type cluster struct {
 	name         string
 	baseURL      string
+	baseURLHost  string
```
Comment (tokers) [inline pkg/apisix/cluster.go:80]: If host is required, we can change the `baseURL` to `pathPrefix`, currently, `baseURLHost` is partial of `baseURL`.

Comment (tokers) [inline pkg/apisix/cluster.go:80]: If host is required, we can change the `baseURL` to `pathPrefix`, currently, `baseURLHost` is partial of `baseURL`.

File: pkg/apisix/cluster.go [hunk 22]
```diff
@@ -290,6 +310,51 @@ func (c *cluster) GlobalRule() GlobalRule {
...
+
+func (c *cluster) healthCheck(ctx context.Context) (err error) {
+	// tcp socket probe
+	d := net.Dialer{Timeout: 3 * time.Second}
+	conn, err := d.DialContext(ctx, "tcp", c.baseURLHost)
+	if err != nil {
+		return err
+	}
+	if er := conn.Close(); er != nil {
+		log.Warnf("failed close tcp socket: %s", er)
```
Comment (tokers) [inline pkg/apisix/cluster.go:353]:
```suggestion
		log.Warnw("failed to close tcp probe connection",
			zap.Error(err),
			zap.String("cluster", c.name),
		)
```

File: pkg/apisix/cluster.go [hunk 23]
```diff
@@ -290,6 +310,51 @@ func (c *cluster) GlobalRule() GlobalRule {
 	return c.globalRules
 }
 
+// HealthCheck implements Cluster.HealthCheck method.
+func (c *cluster) HealthCheck(ctx context.Context) (err error) {
+	if c.cacheSyncErr != nil {
```
Comment (tokers) [inline pkg/apisix/cluster.go:315]: Actually, I think health check should be independent with the first time cache sync.

File: pkg/ingress/controller.go [hunk 24]
```diff
@@ -423,3 +455,21 @@ func (c *Controller) syncSSL(ctx context.Context, ssl *apisixv1.Ssl, event types
 	}
 	return err
 }
+
+func (c *Controller) checkClusterHealth(ctx context.Context, cancelFunc context.CancelFunc) {
+	defer cancelFunc()
+	for {
+		select {
+		case <-ctx.Done():
```
Comment (tokers) [inline pkg/ingress/controller.go:463]: Just return in this arm.

File: pkg/ingress/controller.go [hunk 25]
```diff
@@ -317,30 +325,47 @@ election:
 }
 
 func (c *Controller) run(ctx context.Context) {
-	log.Infow("controller now is running as leader",
+	log.Infow("controller tries to leading ...",
 		zap.String("namespace", c.namespace),
 		zap.String("pod", c.name),
 	)
+
+	var cancelFunc context.CancelFunc
```
Comment (tokers) [inline pkg/ingress/controller.go:333]: It seems that you don't merge from the latest master, here we already have some methods to finalize the context passed to `run`.

Here we cannot cancel the context of `ctx` in parameters, so we cannot give up the leader role by ourselves.

File: pkg/apisix/cluster.go (hunk 21 above)
Comment (slene) [reply pkg/apisix/cluster.go:80]: Split baseURL to pathPrefix and baseURLHost is not enough. Also need scheme. So how to change this config ?

File: pkg/ingress/controller.go (hunk 24 above)
Comment (slene) [reply pkg/ingress/controller.go:463]: ok

File: pkg/ingress/controller.go (hunk 25 above)
Comment (slene) [reply pkg/ingress/controller.go:333]:
```
...
election:
	curCtx, cancel := context.WithCancel(rootCtx)
	c.leaderContextCancelFunc = cancel
	elector.Run(curCtx)
	select {
	case <-rootCtx.Done():
		return nil
	default:
		goto election
	}
...
```
```
func (c *Controller) run(ctx context.Context) {
...
	defer c.leaderContextCancelFunc()
...
	<-ctx.Done()
	c.wg.Wait()
}
```

Not, you mean use one c.leaderContextCancelFunc to control the run context ? It is not enough.
I think the code should wait for all goroutine to exit in c.run before executing the next election.

File: pkg/apisix/cluster.go (hunk 23 above)
Comment (slene) [reply pkg/apisix/cluster.go:315]: You mean health check should wait first time cache sync finished ?

Comment (slene) [reply pkg/apisix/cluster.go:315]:
```
func (c *cluster) HealthCheck(ctx context.Context) (err error) {
	if c.cacheSyncErr != nil {
		err = c.cacheSyncErr
		return
	}
        // _cacheSyncing means the first time cache sync
	if atomic.LoadInt32(&c.cacheState) == _cacheSyncing { 
		return
	}
...
```

File: pkg/apisix/cluster.go (hunk 21 above)
Comment (tokers) [reply pkg/apisix/cluster.go:80]: Or you can just save the `url.URL` object.

File: pkg/apisix/cluster.go (hunk 23 above)
Comment (tokers) [reply pkg/apisix/cluster.go:315]: Yep.

File: pkg/ingress/controller.go (hunk 16 above)
Comment (gxthrj) [reply pkg/ingress/controller.go]: Why mark this as resolved?  in `case <-ctx.Done()` do not need the logic of health check.

File: pkg/ingress/controller.go (hunk 25 above)
Comment (tokers) [reply pkg/ingress/controller.go:333]: The reason why I introduced the `leaderContextCancelFunc` is, package `leaderelection` doesn't quit the lease renew loop even if `c.run` exits.

> Not, you mean use one c.leaderContextCancelFunc to control the run context ? It is not enough.
I think the code should wait for all goroutine to exit in c.run before executing the next election.

Yep, waiting is required before it exits. This is logical inside `c.run`.

File: pkg/apisix/cluster.go (hunk 21 above)
Comment (slene) [reply pkg/apisix/cluster.go:80]: Replace baseURL with *url.URL object may be make more changes. Can we do that in this pr ? Or just use baseURLHost.

**Comment by tokers** [comment]
@slene Any new updates?

File: pkg/apisix/cluster.go (hunk 23 above)
Comment (slene) [reply pkg/apisix/cluster.go:315]: > You mean health check should wait first time cache sync finished ?

The code above already implement it.

- The cluster sync have two state `syncing` `synced` only.
- When state is `synced` and has an error of health check. Just give up leader and re-create cluster to syncing.

File: pkg/ingress/controller.go (hunk 25 above)
Comment (slene) [reply pkg/ingress/controller.go:333]: The cancelFunc manage context of c.run, leaderContextCancelFun just manage the election context.
If call leaderContextCancelFun in health check the election will run into next loop and not wait last c.run exited.

**Comment by slene** [comment]
@tokers no

File: pkg/apisix/cluster.go (hunk 21 above)
Comment (tokers) [reply pkg/apisix/cluster.go:80]: OK, agree with you :), we can change them in subsequent PRs.

File: pkg/ingress/controller.go (hunk 25 above)
Comment (tokers) [reply pkg/ingress/controller.go:333]: I didn't say we should call `leaderContextCancelFun` in the health check. Just call it before the `c.run` return. Or the leader role cannot be given up.

**Review by gxthrj — APPROVED** [review]

**Comment by tokers** [comment]
I have merged your branch with the latest master branch, now LGTM to me.

**Review by tokers — APPROVED** [review]

**Comment by tokers** [comment]
https://github.com/apache/apisix-ingress-controller/pull/499 See this @slene , Corrupted this branch by accident 😂, but don't worry you are still the contributor.

**Comment by slene** [comment]
@tokers Bad news. You performed an incorrect git operation after force push. So my collaboration record is lost. 😞😔😣

**Comment by tokers** [comment]
> @tokers Bad news. You performed an incorrect git operation after force push. So my collaboration record is lost. 😞😔😣

It's may bad. But don't worry your collaboration is still counted, and now you are the contributor.

**Pull request closed without merging**